### ISSUE TITLE
Line breaks does not work with Cairo

### ISSUE THREAD

**Issue by Finnerale** [linebender/piet#169]
While trying to write a multi-line label widget for druid I noticed that line breaks work fine on Windows but not at all on Linux.

Behavior on Windows:
- Breaks lines on whitespace as needed
- `\n` always causes a break

Behavior on Linux:
- No breaks at all

Here is an example as an interactive druid application of this behavior:
```rust
// [dependencies.druid]
// git = "https://github.com/xi-editor/druid"

use druid::{
    piet::{FontBuilder, PietTextLayout, RenderContext, Text, TextLayout, TextLayoutBuilder},
    widget::prelude::*,
    AppLauncher, Point, WidgetExt, WindowDesc,
};

// a fudgey way to get an approximate line height from a font size
const LINE_HEIGHT_FACTOR: f64 = 1.2;
// a fudgey way of figuring out where to put the baseline, relative to line height
const BASELINE_GUESS_FACTOR: f64 = 0.8;
// added padding between the edges of the widget and the text.
const LABEL_X_PADDING: f64 = 2.0;

struct WrapLabel {
    layout: Option<PietTextLayout>,
}

impl WrapLabel {
    pub fn new() -> Self {
        WrapLabel { layout: None }
    }
}

impl Widget<String> for WrapLabel {
    fn event(&mut self, _ctx: &mut EventCtx, _event: &Event, _data: &mut String, _env: &Env) {}
    fn lifecycle(
        &mut self,
        _ctx: &mut LifeCycleCtx,
        _event: &LifeCycle,
        _data: &String,
        _env: &Env,
    ) {
    }
    fn update(&mut self, _ctx: &mut UpdateCtx, _old_data: &String, _data: &String, _env: &Env) {}

    fn layout(
        &mut self,
        ctx: &mut LayoutCtx,
        bc: &BoxConstraints,
        data: &String,
        env: &Env,
    ) -> Size {
        let font_name = env.get(druid::theme::FONT_NAME);
        let font_size = env.get(druid::theme::TEXT_SIZE_NORMAL);
        let font = ctx
            .text()
            .new_font_by_name(font_name, font_size)
            .build()
            .unwrap();

        let layout = ctx
            .text()
            .new_text_layout(&font, &data, bc.max().width)
            .build()
            .unwrap();

        let width = layout.width();
        self.layout = Some(layout);

        bc.constrain(Size::new(width, font_size * LINE_HEIGHT_FACTOR))
    }

    fn paint(&mut self, ctx: &mut PaintCtx, _data: &String, env: &Env) {
        let font_size = env.get(druid::theme::TEXT_SIZE_NORMAL);
        let text_layout = self.layout.as_ref().unwrap();
        let line_height = font_size * LINE_HEIGHT_FACTOR;

        let origin = Point::new(LABEL_X_PADDING, line_height * BASELINE_GUESS_FACTOR);
        let color = env.get(druid::theme::LABEL_COLOR);

        ctx.draw_text(&text_layout, origin, &color);
    }
}

fn make_widget() -> impl Widget<String> {
    WrapLabel::new().center()
}

fn main() {
    let window = WindowDesc::new(make_widget).window_size((400.0, 400.0));
    let state =
        "This is some long text which should cause the WrapLabel to display multiple lines.\nThis one particualarly.".into();
    AppLauncher::with_window(window)
        .launch(state)
        .expect("launch failed");
}
```

And here one using only piet:
```rust
// [dependencies]
// piet = { git = "https://github.com/linebender/piet" }
// piet-common = { git = "https://github.com/linebender/piet", package = "piet-common", features = 

["png"]}
use piet::kurbo::Point;
use piet::{Color, RenderContext, Text, FontBuilder, TextLayoutBuilder};
use piet_common::Device;

// a fudgey way to get an approximate line height from a font size
const LINE_HEIGHT_FACTOR: f64 = 1.2;
// a fudgey way of figuring out where to put the baseline, relative to line height
const BASELINE_GUESS_FACTOR: f64 = 0.8;
// added padding between the edges of the widget and the text.
const LABEL_X_PADDING: f64 = 2.0;

/// Feature "png" needed for save_to_file() and it's disabled by default for optional dependencies
/// cargo run --example png --features png
fn main() {
    let text: String =
        "This is some long text which should cause the WrapLabel to display multiple lines.\nThis one particualarly.".into();

    let mut device = Device::new().unwrap();
    let width = 400;
    let height = 400;
    let mut bitmap = device.bitmap_target(width, height, 1.0).unwrap();
    let mut rc = bitmap.render_context();
    rc.clear(Color::WHITE);


    let font_name = "Arial";
    let font_size = 14.0;
    let font = rc
            .text()
            .new_font_by_name(font_name, font_size)
            .build()
            .unwrap();
    let layout = rc
            .text()
            .new_text_layout(&font, &text, 400.0)
            .build()
            .unwrap();
    let line_height = font_size * LINE_HEIGHT_FACTOR;

    let origin = Point::new(LABEL_X_PADDING, line_height * BASELINE_GUESS_FACTOR);
    let color = Color::rgb(0.0, 0.0, 0.0);

    rc.draw_text(&layout, origin, &color);


    rc.finish().unwrap();
    std::mem::drop(rc);

    bitmap
        .save_to_file("temp-image.png")
        .expect("file save error");
}
```

**Comment by hwchen** [comment]
Ah, I see you're using `draw_text`. I had anticipated that users of piet would use deal with multiline more manually, for example to deal with scrolling and rendering only parts of a text layout. 

I could be wrong; but in any case `draw_text` should be consistent across platforms.

@cmyr and @raphlinus some guidance would be helpful here. Should I just implement `draw_text` for cairo and web, where all lines are rendered?

If I remember correctly, there was a bit of a conundrum, where I could render just a range in cairo, but it was a bit trickier in d2d.

**Comment by raphlinus** [comment]
Yes, implement multiline `draw_text`.

I think there will be both simpler and more sophisticated uses of text, the latter of which may do their own processing with ranges.

**Comment by hwchen** [comment]
Closed by #170
